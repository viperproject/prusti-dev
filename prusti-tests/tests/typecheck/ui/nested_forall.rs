// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Pno_verify=true -Phide_uuids=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "\[[a-z0-9]{4}\]::" -> "[$(CRATE_ID)]::"

use prusti_contracts::*;

#[requires(forall(|a: i32| forall(|a: i32| a==a)))]
fn test1() {}

#[requires(forall(|a: i32| forall(|b: i32| a==a ==> b==b)))]
fn test2() {}

#[requires(forall(|a: i32| forall(|b: i32| forall(|c: i32| a==a && b==b))))]
fn test3() {}

#[requires(exists(|a: i32| exists(|a: i32| a==a)))]
fn test4() {}

#[requires(exists(|a: i32| exists(|b: i32| a==a ==> b==b)))]
fn test5() {}

#[requires(exists(|a: i32| exists(|b: i32| exists(|c: i32| a==a && b==b))))]
fn test6() {}


fn main() {}
