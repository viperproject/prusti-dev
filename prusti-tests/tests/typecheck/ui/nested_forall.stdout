// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Pno_verify=true -Phide_uuids=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "/[[a-z0-9]{4}/]::" -> "[$(CRATE_ID)]::"









#![feature(register_tool)]
#![register_tool(prusti)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
use prusti_contracts::*;
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}},[]]}},[]]}}"]
fn prusti_pre_item_test1_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |a: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    || -> bool { a == a };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test1() { }
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":104}}}]}},[]]}},[]]}}"]
fn prusti_pre_item_test2_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |b: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    || -> bool { a == a };

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_104"]
                    || -> bool { b == b };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test2() { }
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"ForAll/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":103,/"count/":1},{/"kind/":{/"And/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":104}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":105}}}]}},[]]}},[]]}},[]]}}"]
fn prusti_pre_item_test3_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |b: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    |c: i32|
                        {

                            #[prusti::spec_only]
                            #[prusti::expr_id =
                              "$(NUM_UUID)_104"]
                            || -> bool { a == a };

                            #[prusti::spec_only]
                            #[prusti::expr_id =
                              "$(NUM_UUID)_105"]
                            || -> bool { b == b };
                        };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test3() { }
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}},[]]}},[]]}}"]
fn prusti_pre_item_test4_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |a: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    || -> bool { a == a };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test4() { }
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":104}}}]}},[]]}},[]]}}"]
fn prusti_pre_item_test5_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |b: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    || -> bool { a == a };

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_104"]
                    || -> bool { b == b };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test5() { }
#[allow(unused_must_use, unused_variables, dead_code)]
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":101,/"count/":1},{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":102,/"count/":1},{/"kind/":{/"Exists/":[{/"spec_id/":/"$(UUID)/",/"expr_id/":103,/"count/":1},{/"kind/":{/"And/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":104}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":105}}}]}},[]]}},[]]}},[]]}}"]
fn prusti_pre_item_test6_$(NUM_UUID)() {

    #[prusti::spec_only]
    #[prusti::expr_id = "$(NUM_UUID)_101"]
    |a: i32|
        {

            #[prusti::spec_only]
            #[prusti::expr_id = "$(NUM_UUID)_102"]
            |b: i32|
                {

                    #[prusti::spec_only]
                    #[prusti::expr_id =
                      "$(NUM_UUID)_103"]
                    |c: i32|
                        {

                            #[prusti::spec_only]
                            #[prusti::expr_id =
                              "$(NUM_UUID)_104"]
                            || -> bool { a == a };

                            #[prusti::spec_only]
                            #[prusti::expr_id =
                              "$(NUM_UUID)_105"]
                            || -> bool { b == b };
                        };
                };
        };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test6() { }
fn main() { }
Procedure(ProcedureSpecification { pres: [Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: DefId(0:8 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test1_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: DefId(0:26 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test4_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: DefId(0:31 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test5_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}) }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(104), expr: DefId(0:32 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test5_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#1}) }) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
Procedure(ProcedureSpecification { pres: [Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: DefId(0:13 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test2_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}) }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(104), expr: DefId(0:14 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test2_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#1}) }) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: Exists(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: And([Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(104), expr: DefId(0:38 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test6_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}::{closure#0}) }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(105), expr: DefId(0:39 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test6_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}::{closure#1}) }) }]) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
Procedure(ProcedureSpecification { pres: [Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: ForAll(QuantifierVars { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), vars: [(_2, i32)] }, TriggerSet([]), Assertion { kind: And([Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(104), expr: DefId(0:20 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test3_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}::{closure#0}) }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(105), expr: DefId(0:21 ~ nested_forall[$(CRATE_ID)]::prusti_pre_item_test3_$(NUM_UUID)::{closure#0}::{closure#0}::{closure#0}::{closure#1}) }) }]) }) }) }) }], posts: [], pledges: [], predicate_body: None, pure: false, trusted: false })
